code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingExtension.java

```diff
@@ -46,6 +48,9 @@ public final class MessagingExtension extends Extension {
     final ContentCardRulesEngine contentCardRulesEngine;
     private SerialWorkDispatcher<Event> serialWorkDispatcher;
 
+    private static final Object completionHandlersMutex = new Object();
+    private static List<CompletionHandler> completionHandlers = new ArrayList<>();
```

Comment:
Wondering if it would be better for performance to use `ConcurrentLinkedQueue` here instead of a regular ArrayList with `synchronized` blocked for reads and writes. Thoughts?

Reply:
Discussed offline. Since this API only gets called when the push notification is clicked, we don't expect it to get be called frequently. Leaving this as is to revisit if we face performance issues.